provide average data usage per timespan for SIMs

Lenakh97 opened the pull request.

Lenakh97 committed:
feat: provide data usage per timeSpan in details

coderbyheart commented on lambda/getDataUsagePerTimespan.spec.ts:
```diff
@@ -0,0 +1,36 @@
+			lastMonth: 75,
+		}
+		const getUsage = getDataUsagePerTimespan({
+			getBinInterval: async () =>
+				Promise.resolve([
+					{
+						ICCID: '89444612812874751710',
+						ID: '4a4d299cffa7c4096f1577881a340cc32f264c12',
+						measure_name: 'UsedBytes',
+						time: new Date('2024-07-15T12:05:58.547Z'),
+						'measure_value::double': 25,
+					},
+					{
+						ICCID: '89444612812874751710',
+						ID: '9b3e92e90836e08aad8ea4f408dae6006e817506',
+						measure_name: 'UsedBytes',
+						time: new Date('2024-07-15T12:40:38.204Z'),
+						'measure_value::double': 50,
+					},
+				]),
```
Here is how to mock a function that returns different values based on the argument.

```suggestion
			getBinInterval: async (interval) => {
			  switch (interval.durationHours) {
			    case 1:
			      return Promise.resolve([
					{
						ICCID: '89444612812874751710',
						ID: '4a4d299cffa7c4096f1577881a340cc32f264c12',
						measure_name: 'UsedBytes',
						time: new Date('2024-07-15T12:05:58.547Z'),
						'measure_value::double': 25,
					},
					{
						ICCID: '89444612812874751710',
						ID: '9b3e92e90836e08aad8ea4f408dae6006e817506',
						measure_name: 'UsedBytes',
						time: new Date('2024-07-15T12:40:38.204Z'),
						'measure_value::double': 50,
					},
				  ])
			    default: throw new Error(`Unknown interval!`)
			  }  
			}
```

Lenakh97 commented:
I have added the test as you suggested above! I saw that I'm using getBinInterval() two places, and have kept the function for now. Both because it is used two places and because I based the test 'getDataUsagePerTimeSpan.spec.ts' on this function. Is this okay or could I maybe pass in the 'getQueryString()' function in stead to get the correct timespan and base the test on that function?
I have also added a test for the Querystring.

coderbyheart commented on lambda/getBinInterval.ts:
```diff
@@ -15,14 +16,13 @@ export const getBinInterval =
 		durationHours: number
 		iccid: string
 	}): Promise<Record<string, unknown>[]> => {
```
**problem:** Do not use `unknown` here. You know exactly the type of the returned result.

coderbyheart commented on lambda/getBinInterval.ts:
```diff
@@ -15,14 +16,13 @@ export const getBinInterval =
 		durationHours: number
```
**problem:** use the type `HistoricalDataTimeSpan` here so you ensure that these property names align across the codebase (and it's clear where they are coming from)

coderbyheart commented on lambda/getBasicInformationLambda.ts:
```diff
@@ -110,7 +115,11 @@ export const handler = async (
 			expires: 300,
 		})({ measurements })
 	}
+	const dataUsagePerTimespan = await getDataUsagePerTimespanFunc(iccid)
```
**refactor**: Inline this function here. It then also becomes more clear what the data structure is that you are are returning.

Lenakh97 replied:
That means that we don't need a test for that function right?

coderbyheart replied:
No, it's 4 calls to the the function that creates the query (which is tested) and then calls to functions which are not part of our source, so an extra test is not needed.

coderbyheart replied:
No, as in, we do not need a test.

coderbyheart commented on lambda/getBinInterval.ts:
```diff
@@ -3,6 +3,7 @@ import {
 import { parseResult } from '@bifravst/timestream-helpers'
+import { getQueryString } from './getQueryString.js'
 
 export const getBinInterval =
```
**problem:** The function name is misleading. It is not return a bin interval, but a list of records for a given interval.

coderbyheart commented on lambda/getQueryString.ts:
```diff
@@ -0,0 +1,21 @@
+export const getQueryString = ({
+	binIntervalMinutes,
```
**problem:** use the type HistoricalDataTimeSpan here so you ensure that these property names align across the codebase (and it's clear where they are coming from)

Lenakh97 replied:
Can I make the 'expiredMinutes' property optional to use this type? I'm not using it anywhere now.

coderbyheart replied:
Yes, you can use for example `Pick<HistoricalDataTimeSpan, 'binIntervalMinutes' | 'durationHours'>` to construct a new type based on `HistoricalDataTimeSpan` that only has the two required properties.

coderbyheart replied:
https://www.typescriptlang.org/docs/handbook/utility-types.html#picktype-keys

coderbyheart commented on cdk/BackendStack.ts:
```diff
@@ -61,10 +61,10 @@ export class BackendStack extends Stack {
 		})
 		const db = new Timestream.CfnDatabase(this, 'db')
-		const table = new Timestream.CfnTable(this, 'table', {
+		const table = new Timestream.CfnTable(this, 'table2', {
```
**problem:** Local testing artifact, revert.

coderbyheart commented on cdk/BackendStack.ts:
```diff
@@ -61,10 +61,10 @@ export class BackendStack extends Stack {
 			databaseName: db.ref,
 			retentionProperties: {
-				MemoryStoreRetentionPeriodInHours: `24`,
+				MemoryStoreRetentionPeriodInHours: `72`,
```
**problem:** Testing artifact. Revert.

Lenakh97 replied:
This would make the test fail in CI? Should I make sure the retentionPeriod is 72 hours only in CI?

coderbyheart replied:
Yes, example: https://github.com/hello-nrfcloud/backend/blob/282cc328e5d2469a3326b0e1e906985b43e39a44/cdk/resources/DeviceStorage.ts#L16

coderbyheart commented on e2e-tests/e2eTest.spec.ts:
```diff
@@ -76,21 +85,55 @@ void describe('e2e-tests', () => {
+			lastWeek: 45,
+		},
 	}
 	for (const [iccid, response, statusCode, status] of [
```
Refactor the test case so it reads:

```
seed
query
assert
```

so the data in question is close together and can be inspected at the same time without need to jump back and forth in the file.

Lenakh97 committed:
fix: don't put 0 values for usage data into db

coderbyheart commented on lambda/listRecordsForInterval.ts:
```diff
@@ -0,0 +1,50 @@
+				}),
+			)
+		} catch (err) {
+			if (err instanceof ValidationException) {
```
This would shadow an actual error in your SQL, so instead of catching this, make sure to only query the available columns.

You can get the available columns by doing a `SELECT * FROM table LIMIT 1`, and through that determine which columns are present.

See https://github.com/hello-nrfcloud/backend/blob/19a9e47d54cbc7404017e1bd335570cad8c25f66/historicalData/getAvailableColumns.ts#L6
and https://github.com/hello-nrfcloud/backend/blob/19a9e47d54cbc7404017e1bd335570cad8c25f66/lambda/queryLwM2MHistory.ts#L122-L133

Lenakh97 replied:
I'm not sure this solves my use case? In my use case the problem is that the ICCID is not in Timestream. This would probably be smart to implement anyway - but I believe I would need another solution for my initial problem. In the  use case I have been looking at the Timestream table exists and there is data in there meaning that the columns exists. The only problem is that there is no data for my exact ICCID, and that results in a ValidationException. Any suggestions?

coderbyheart replied:
As discussed: a `select * from table where ICCD='nonexistent'` will not yield a `ValidationException` but return an empty result.

Make sure that the column exists that you are querying (because you're not using `SELECT *`).

Lenakh97 committed:
refactor: refactor tests and add comments

coderbyheart commented on e2e-tests/e2eTest.spec.ts:
```diff
@@ -25,88 +25,193 @@ const iccidOld = getRandomICCID(4573)
-			randomUsedBytesVal,
-			iccidNewWL,
+		await seedTimestreamFunc({
+			usage: usageLastHour,
```
*problem:* Please don't use magic here, seed the database with the actual values, make them inline:

```typescript
await seedDbWith([
{value: 67, ts: Date}
])
```

so it becomes clear which data is present in the DB, without having to jump back and forth.

Lenakh97 replied:
I understand that this would make it easier to read, but I don't understand how I can make the actual values (ts) inline? I need to use the ts for 'lastHour', 'lastDay' and 'lastTwoDays'. This can not be hardcoded; and I need to check it later in the e2e test meaning I need to make it a const outside of the test case? Or am I missing something?

coderbyheart replied:
You still can initialize these values in the test case:

```typescript

// Setup
const lastHour = getLastHourDate()
await seedDb([
{value: 67, ts: lastHour}
])

// Execute
const result = await query(...)

// Assert
...
```

coderbyheart replied:
I've looked at the test and yes, it needs a bit of refactoring to navigate out of this setup.

I've extracted the first e2e test in its own file and made it follow the structure that I've outlined above: https://github.com/bifravst/sim-details/blob/973c22b8bb218cc278cb710652b8fc4d60e2630b/e2e-tests/getSIMDetails.spec.ts#L32

coderbyheart committed:
refactor: extract test (#193)

* refactor: extract test

Make the test set-up isolated and do not use magic
to construct the seed data.

This makes the seeded data explicit, but still
random.

* ci: this is not applicable in this project

* test(e2e): run all e2e tests

* test(e2e): separate tests

It is better to not test for incorrect data,
instead only test what should work.

* refactor(e2e): move shared clients to function

* test(e2e): mark as todo

This no longer works because it depended on the first
test to run

* refactor(e2e): extract test cases to individual files

---------

Co-authored-by: Lena Haraldseid <lena.haraldseid@nordicsemi.no>

Lenakh97 committed:
refactor: remove e2eTest.spec.ts
